halo2-base/README.md

```diff
@@ -0,0 +1,498 @@
+    let mut first_pass = SKIP_FIRST_PASS;
+    let mut assigned_advices = HashMap::new();
+    layouter
+        .assign_region(
+            || "GateCircuitBuilder generated circuit",
+            |mut region| {
+                if first_pass {
+                    first_pass = false;
+                    return Ok(());
+                }
+                    
+                if !self.builder.borrow().witness_gen_only {
+                // clone the builder so we can re-use the circuit for both vk and pk gen
+                    let builder = self.builder.borrow().clone();
+                    for threads in builder.threads.iter().skip(1) {
+                        assert!(
+                            threads.is_empty(),
+                            "GateCircuitBuilder only supports FirstPhase for now"
+                        );
+                    }
+                    let assignments = builder.assign_all(
+                            gate,
+                            lookup_advice,
+                            q_lookup,
+                            &mut region,
+                            Default::default(),
+                    );
+                    *self.break_points.borrow_mut() = assignments.break_points;
+                    assigned_advices = assignments.assigned_advices;
+                } else {
+                    // If we are only generating witness, we can skip the first pass and assign threads directly
+                    let builder = self.builder.take();
+                    let break_points = self.break_points.take();
+                    for (phase, (threads, break_points)) in builder
+                        .threads
+                        .into_iter()
+                        .zip(break_points.into_iter())
+                        .enumerate()
+                        .take(1)
+                    {
+                        assign_threads_in(
+                            phase,
+                            threads,
+                            gate,
+                            lookup_advice.get(phase).unwrap_or(&vec![]),
+                            &mut region,
+                            break_points,
+                        );
+                    }
+                }
+                Ok(())
+            },
+        )
+        .unwrap();
+    assigned_advices
+    }
+}
+
+```
+`layouter`'s `assign_region()` function is invoked which yields a mutable reference to `Region`. `region` is used to assign of a contiguous region of cells representing the circuit within Halo2's proving system.
+
+If `witnes_gen_only` is not set within the `builder` (during keygen, and mock proving) `sub_synthesize` takes ownership of the `builder`, asserts all threads but the thread corresponding to the first phase are empty, and calls `assign_all()` to assign all cells within this `thread`'s context to the circuit. The resulting column breakpoints are recorded in `GateCircuitBuilder`'s `break_points` field. 
```

Comment:
probably my fault, but saying `thread` singular to refer to multiple `Context` is a little confusing